compiler/executable/match_/instructions/thing.rs

```diff
@@ -21,6 +21,43 @@ use crate::{
     executable::match_::instructions::{CheckInstruction, DisplayVec, Inputs},
 };
 
+#[derive(Debug, Clone)]
+pub struct IidInstruction<ID> {
+    pub iid: Iid<ID>,
+    pub types: Arc<BTreeSet<Type>>,
+    pub checks: Vec<CheckInstruction<ID>>,
+}
+
+impl IidInstruction<Variable> {
+    pub fn new(iid: Iid<Variable>, type_annotations: &TypeAnnotations) -> Self {
+        let types = type_annotations.vertex_annotations_of(iid.var()).unwrap().clone();
+        Self { iid, types, checks: Vec::new() }
+    }
+}
+
+impl<ID> IidInstruction<ID> {
+    pub(crate) fn add_check(&mut self, check: CheckInstruction<ID>) {
+        self.checks.push(check)
+    }
+}
+
+impl<ID: IrID> IidInstruction<ID> {
+    pub fn map<T: IrID>(self, mapping: &HashMap<ID, T>) -> IidInstruction<T> {
+        let Self { iid, types, checks } = self;
+        IidInstruction {
+            iid: iid.map(mapping),
+            types,
+            checks: checks.into_iter().map(|check| check.map(mapping)).collect(),
+        }
+    }
+}
+
+impl<ID: IrID> fmt::Display for IidInstruction<ID> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "[{}] filter {}", &self.iid, DisplayVec::new(&self.checks))
```

Comment:
shouldn't this mention iid somewhere :)

Reply:
`self.iid` is the iid constraint, which prints `"$1 iid P[IID(0)]` e.g.